### DISCUSSION TITLE
How to inject dependencies into a plugin? Or what is the best practice for a work around?

### DISCUSSION THREAD

**Discussion by sundeepgupta** [ionic-team/capacitor#3923] [Q&A]
I'm looking at building an [iOS plugin](https://capacitorjs.com/docs/plugins/ios), although my question applies to Android as well. What I've noticed is that the plugin is initialized directly by the bridge. And obviously plugin function calls are also made by the bridge. So, it doesn't seem possible to use constructor or parameter injection. Is this correct? If so, I can see a couple of work arounds, but wanted to first make sure I'm not missing something.

**Comment by ikeith** [comment]
Correct. There is no scheme to handle injection for you. The plugin is expected to manage its own dependencies. The bridge will call the [load](https://capacitorjs.com/docs/plugins/ios#adding-initialization-logic) method on each plugin after it is instantiated and before it receives any calls from the application. So that is where it should configure anything that it will need later.

**Reply by sundeepgupta** [reply]
Understood - thanks. It would be great, perhaps for future consideration, if initialization could be (optionally) controlled by the native code to allow for constructor injection.

**Comment by mlostekk** [comment]
A DI mechanism is definitely required for serious usage of capacitor, like the "ModuleFactory" approach in React Native. 

Nobody wants to handle god objects.
